internal/engine/dbfs/dbfs.go

```diff
@@ -0,0 +1,179 @@
+package dbfs
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+	"gopkg.in/yaml.v3"
+
+	"github.com/xqueries/xdb/internal/id"
+)
+
+const defaultPerm os.FileMode = 0600
+
+// DBFS represents the structure of a single DataBase-FileSystem.
+// It provides easy access to all folders and files within the root
+// directory, with wrappers for easily modifying the files.
+type DBFS struct {
+	fs afero.Fs
+}
+
+// CreateNew initializes a new, empty DBFS in the root of the given file system.
+func CreateNew(fs afero.Fs) (*DBFS, error) {
+	dbfs := &DBFS{
+		fs: fs,
+	}
+
+	if err := dbfs.mkdir(TablesDirectory); err != nil {
+		return nil, err
+	}
+	if err := dbfs.touch(filepath.Join(TablesDirectory, TablesInfoFile)); err != nil {
+		return nil, err
+	}
+
+	if err := dbfs.Close(); err != nil {
+		return nil, fmt.Errorf("close stub dbfs: %w", err)
+	}
+	return Load(fs)
+}
+
+// Load loads a DBFS from the given file system.
+func Load(fs afero.Fs) (*DBFS, error) {
+	// FIXME: perform checks
+	return &DBFS{
+		fs: fs,
+	}, nil
+}
+
+// Close closes the DBFS and releases all resources.
+func (dbfs *DBFS) Close() error {
+	// nothing to do here yet
+	return nil
+}
+
+// Table returns a Table object representing the files and directories
+// for the table with the given name, or an error if the table does not
+// exist. Call HasTable to see whether a table with a given name exists.
+func (dbfs *DBFS) Table(name string) (Table, error) {
+	infos, err := dbfs.loadTablesInfo()
+	if err != nil {
+		return Table{}, err
+	}
+
+	// perform checks
+	tblID, ok := infos[name]
+	if !ok {
+		return Table{}, fmt.Errorf("table '%s' does not exist", name)
+	}
+
+	return Table{
+		fs: afero.NewBasePathFs(dbfs.fs, filepath.Join(TablesDirectory, tblID)),
+	}, nil
+}
+
+// HasTable determines whether or not a table with the given name exists in this
+// database.
+func (dbfs *DBFS) HasTable(name string) (bool, error) {
+	infos, err := dbfs.loadTablesInfo()
+	if err != nil {
+		return false, err
+	}
+
+	_, ok := infos[name]
+	return ok, nil
+}
+
+// CreateTable creates empty table files for a table with the given name.
+// This will return an error if a table with the given name already exists.
+// Calling this method will also create an entry in the tables info file.
+func (dbfs *DBFS) CreateTable(name string) (Table, error) {
+	infos, err := dbfs.loadTablesInfo()
```

Comment:
not worth a separate function really, it's just a map access and checks.
I wanted to avoid a `func Table(name string) (Table, bool, error)`, since I really don't like bool and error returned for some reason. Probably because it requires more checks for the caller, which doesn't make calling code prettier.

Reply:
Sure alright